### DISCUSSION TITLE
About support of Xamarin projects by Uno Platform

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#9604] [Q&A] [answered]
[_From Discord:_](https://discord.com/channels/372137812037730304/550416151172087808/1012234951057490010)

Does anyone know how we will handle the end of Xamarin support? Rebasing from Xamarin to .NET MAUI? 
https://dotnet.microsoft.com/en-us/platform/support/policy/xamarin

**Comment by pkar70** [comment]
Please, do not drop current ability to port UWP app to app that compiles for UWP 15063 (last version on phones), and for Android/iOS.

**Comment by jeromelaban** [comment, marked as answer]
By this particular article (as of 2022-08-25):

_Xamarin support will end on May 1, 2024 for all Xamarin SDKs. Android 13 and Xcode 14 SDKs (iOS and iPadOS 16, macOS 13) will be the final versions Xamarin will target._

This means that Uno is going to be providing support for Xamarin (`monoandroid12`, `monoandroid13`, `xamarinios10` and `xamarinmac20`) by that date as well. Note that after that, apps built with these SDK will not be accepted by both apple and google stores.

That being said, Uno now supports running on top of iOS/Android/macOS/Catalyst for .NET 6 (underlying platform support part used by MAUI) meaning that you can already migrate your application to this new target. Now that those .NET pieces are in place, we are in the process of updating our documentation to provide guidance on how to migrate your apps.

In the meantime, in a nutshell, there are two paths to migrate from Uno UWP app based on Xamarin to .NET 6.
## The fastest path to keep using the UWP APIs
   1. In a separate folder, create a new application from the templates using [`dotnet new unoapp-uwp-net6 -o MyApp`](https://platform.uno/docs/articles/get-started-dotnet-new.html) (MyApp being of the same name as your existing app), 
   2. Move the `MyApp.Mobile` project folder and include it in your existing solution
   3. Make the adjustments for your new application head to build
   4. Optionally remove the legacy Xamarin heads, once you're done with the migration

## The forward looking path using the WinUI/WinApp SDK APIs
   1. In a separate folder, create a new application from the templates using [`dotnet new unoapp -o MyApp`](https://platform.uno/docs/articles/get-started-dotnet-new.html) (MyApp being of the same name as your existing app), or by using the Visual Studio `Uno Platform App` template
   2. Move the `MyApp.Mobile` project folder and include it in your existing solution
   3. Make the adjustments to move from the [UWP APIs to the WinUI APIs](https://platform.uno/docs/articles/updating-to-winui3.html)
   4. Optionally [convert your legacy Xamarin heads](https://platform.uno/docs/articles/updating-to-winui3.html) to use Uno.WinUI instead of Uno.UI

Note that since .NET 6 breaks binary compatibility with Xamarin, most of the existing nuget packages that target `monoandroidXX`, `xamarinios10` and `xamarinmac20` (bindings to native SDKs are a good example) will not work properly and will need an updated version that are compatible with `net6.0-ios`, `net6.0-android`, `net6.0-maccatalyst` and `net6.0-macos`.

**Reply by crogeniks** [reply]
> Note that after that, apps built with these SDK will not be accepted by both apple and google stores.

Is there an article with more details about this specific detail?

**Reply by jeromelaban** [reply]
I don't think there's one specifically for what's going to happen in late 2024, but [this one ](https://developer.android.com/google/play/requirements/target-sdk)can be extrapolated for Android: 

> _New apps must target Android 12 (API level 31) or higher; except for [Wear OS](https://developer.android.com/wear) apps, which must target Android 11 (API level 30) or higher.  Starting in November 2022, app updates must target API level 31 or above and adjust for behavioral changes in Android 12; except for Wear OS apps, which must target API level 30 or higher._

For iOS, [this article](https://developer.apple.com/support/xcode/) details it:
> _Minimum SDK requirements for App Store submissions: iOS and iPadOS apps submitted to the App Store must be built with Xcode 13 and the iOS 15 SDK. And watchOS apps submitted to the App Store must be built with Xcode 13 and the watchOS 8 SDK._

As far as Visual Studio is concerned, if we can extrapolate a bit, this means that as of late 2024, we're talking about Android 14 as a minimum version, and iOS 17 or 18 as minimum version, which are not going to be supported by Visual Studio at that time.

**Comment by pkar70** [comment]
MAUI nor .Net 6 apps would work on UWP 15063. So, if someone want to make app available for phones, one should compile it as "plain" UWP (outside Uno) and again within Uno for access newer OS.
It (partly) can be done, moving most of app code to Class Library based on .Net Standard 1.4; but most time many lines of code would be required to be duplicated between "plain UWP" and "Uno".
